File: ruby/aoc/lib/aoc/day06b.rb
module AOC
  class Day06b < Day
    def winning_strategies
      time, distance = parse_input
      (1..time).select do |hold|
        hold * (time - hold) > distance
      end.count
    end

    def parse_input
      time, distance = File.readlines(@path, chomp: true)
      _, time = time.split(': ')
      time = time.gsub(" ", "").to_i
      _, distance = distance.split(': ')
      distance = distance.gsub(" ", "").to_i
      [time, distance]
    end
  end
end

File: elixir/aoc/lib/aoc/day06b.ex
defmodule AOC.Day06b do
  use AOC

  def solution(path) do
    [time, distance] = path |> read_lines() |> Enum.map(&String.trim/1)
    [_, time] = time |> String.split(":", trim: true)
    time = time |> String.replace(" ", "", global: true) |> String.to_integer()
    [_, distance] = distance |> String.split(":", trim: true)
    distance = distance |> String.replace(" ", "", global: true) |> String.to_integer()
    winning_strategies(time, distance)
  end

  def winning_strategies(time, distance) do
    1..time
    |> Enum.filter(&winning?(&1, time, distance))
    |> Enum.count()
  end

  def winning?(holding, time, distance) do
    holding * (time - holding) > distance
  end
end
